cluster/mesos/cluster.go

```diff
@@ -0,0 +1,485 @@
+		}
+	}
+
+	return info
+}
+
+func (c *Cluster) addOffer(offer *mesosproto.Offer) {
+	s, ok := c.slaves[offer.SlaveId.GetValue()]
+	if !ok {
+		return
+	}
+	s.addOffer(offer)
+	go func(offer *mesosproto.Offer) {
+		<-time.After(c.offerTimeout)
+		if c.removeOffer(offer) {
+			if _, err := c.driver.DeclineOffer(offer.Id, &mesosproto.Filters{}); err != nil {
+				log.WithFields(log.Fields{"name": "mesos"}).Errorf("Error while declining offer %q: %v", offer.Id.GetValue(), err)
+			} else {
+				log.WithFields(log.Fields{"name": "mesos"}).Debugf("Offer %q declined successfully", offer.Id.GetValue())
+			}
+		}
+	}(offer)
+}
+
+func (c *Cluster) removeOffer(offer *mesosproto.Offer) bool {
+	log.WithFields(log.Fields{"name": "mesos", "offerID": offer.Id.String()}).Debug("Removing offer")
+	s, ok := c.slaves[offer.SlaveId.GetValue()]
+	if !ok {
+		return false
+	}
+	found := s.removeOffer(offer.Id.GetValue())
+	if s.empty() {
+		// Disconnect from engine
+		delete(c.slaves, offer.SlaveId.GetValue())
+	}
+	return found
+}
+
+func (c *Cluster) scheduleTask(t *task) bool {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
+
+	n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), t.config)
+	if err != nil {
+		return false
+	}
+	s, ok := c.slaves[n.ID]
+	if !ok {
+		t.error <- fmt.Errorf("Unable to create on slave %q", n.ID)
+		return true
+	}
+
+	// build the offer from it's internal config and set the slaveID
+	t.build(n.ID)
+
+	c.Lock()
+	// TODO: Only use the offer we need
+	offerIds := []*mesosproto.OfferID{}
+	for _, offer := range c.slaves[n.ID].offers {
+		offerIds = append(offerIds, offer.Id)
+	}
+
+	if _, err := c.driver.LaunchTasks(offerIds, []*mesosproto.TaskInfo{&t.TaskInfo}, &mesosproto.Filters{}); err != nil {
+		// TODO: Do not erase all the offers, only the one used
+		for _, offer := range s.offers {
+			c.removeOffer(offer)
+		}
+		s.Unlock()
+		t.error <- err
+		return true
+	}
+
+	s.addTask(t)
+
+	// TODO: Do not erase all the offers, only the one used
+	for _, offer := range s.offers {
+		c.removeOffer(offer)
+	}
+	c.Unlock()
+	// block until we get the container
+	finished, err := t.monitor()
+
+	if err != nil {
+		//remove task
+		s.removeTask(t.TaskInfo.TaskId.GetValue())
+		t.error <- err
+		return true
+	}
+	if !finished {
+		go func() {
+			for {
+				finished, err := t.monitor()
+				if err != nil {
+					// TODO proper error message
+					log.Error(err)
+					break
+				}
+				if finished {
+					break
+				}
+			}
+			//remove task
+		}()
+	}
+
+	// Register the container immediately while waiting for a state refresh.
+	// Force a state refresh to pick up the newly created container.
+	// FIXME: unexport this method, see FIXME in engine.go
+	s.engine.RefreshContainers(true)
+
+	// TODO: We have to return the right container that was just created.
+	// Once we receive the ContainerID from the executor.
+	for _, container := range s.engine.Containers() {
```

Comment:
I see!

Well, in case the Mesos change doesn't arrive on time (or actually, even to make Swarm work with the latest mesos release without requiring master), what if we injected a label into the container then use said label to find our container?